File: leetcode/maxSubArray.ts
function maxSubArray(nums: number[]): number {
  let max = nums[0];
  let curr = max;
  for (let i: number = 1; i < nums.length; i++) {
    curr += nums[i];

    if (nums[i] > curr) {
      max = Math.max(nums[i], max);
      curr = nums[i];
    } else {
      console.log(max, curr);
      max = Math.max(max, curr);
    }
  }
  return max;
}

const nums = [0, -2, -3];

console.log(maxSubArray(nums));

File: 2025/leetcode /maxSubArray.py
from typing import List

# Also to research the Bruteforce way to solving this

#  to write about
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
      max_count = nums[0]
      count = nums[0]
      i = 1
      arr_len = len(nums)

      while i < arr_len:
        count  = count + nums[i]
        if count < nums[i]:
          count = nums[i]
        
        max_count = max(max_count, count)
        i += 1

      return max_count


sol = Solution()
nums = [5,4,-1,7,8]
print(sol.maxSubArray(nums))